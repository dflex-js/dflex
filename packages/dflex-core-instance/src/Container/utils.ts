import { getElmOverflow, PointBool, Axis } from "@dflex/utils";

const OVERFLOW_REGEX = /(auto|scroll|overlay)/;

type CSSOverflow = ReturnType<typeof getElmOverflow>;

function hasScrollableContent(axis: Axis, DOM: HTMLElement): boolean {
  return axis === "y"
    ? DOM.scrollHeight > DOM.clientHeight
    : DOM.scrollWidth > DOM.clientWidth;
}

function hasOverflowProperty(axis: Axis, DOM: HTMLElement): boolean {
  const overflow: CSSOverflow = getElmOverflow(
    DOM,
    axis === "x" ? "overflow-x" : "overflow-y",
  );

  return OVERFLOW_REGEX.test(overflow);
}

function hasOverflow(axis: Axis, DOM: HTMLElement): boolean {
  if (hasOverflowProperty(axis, DOM)) {
    return hasScrollableContent(axis, DOM);
  }

  return false;
}
