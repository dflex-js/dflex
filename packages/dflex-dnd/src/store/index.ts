// eslint-disable-next-line import/prefer-default-export
export { default as exportedStoreSingleton } from "./exportedStoreSingleton";
export { default as store } from "./storeSingleton";
