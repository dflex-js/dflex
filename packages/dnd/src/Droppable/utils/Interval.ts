/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

class Interval {
  intervalID: number | null;

  constructor() {
    this.intervalID = 0;
  }

  set(cb: Function, delay: number) {
    this.intervalID = setInterval(cb, delay);
  }

  clear() {
    clearInterval(this.intervalID!);
    this.intervalID = null;
  }
}

export default Interval;
