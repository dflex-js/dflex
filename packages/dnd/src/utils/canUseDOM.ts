/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

function canUseDOM() {
  return (
    typeof window !== "undefined" &&
    typeof window.document !== "undefined" &&
    typeof window.document.createElement !== "undefined"
  );
}

export default canUseDOM;
