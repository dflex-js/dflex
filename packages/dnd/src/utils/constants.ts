import type { DefaultDndOpts } from "../types";

export function noop() {}

export const defaultOpts: DefaultDndOpts = Object.freeze({
  enableContainersTransition: true,

  threshold: {
    vertical: 60,
    horizontal: 60,
  },

  restrictions: {
    self: {
      allowLeavingFromTop: true,
      allowLeavingFromBottom: true,
      allowLeavingFromLeft: true,
      allowLeavingFromRight: true,
    },
    container: {
      allowLeavingFromTop: true,
      allowLeavingFromBottom: true,
      allowLeavingFromLeft: true,
      allowLeavingFromRight: true,
    },
  },

  scroll: {
    enable: true,
    initialSpeed: 10,
    threshold: {
      vertical: 15,
      horizontal: 15,
    },
  },

  events: {
    onDragOver: noop,
    onDragLeave: noop,
    onDragOutContainer: noop,
    onDragOutThreshold: noop,
    onLiftUpSiblings: noop,
    onMoveDownSiblings: noop,
    onStateChange: noop,
  },
});
