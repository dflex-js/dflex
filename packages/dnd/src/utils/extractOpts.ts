/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

// @ts-nocheck

// TODO: Do you love typescript? Here we go buddy. It's your chance. Start your
// PR. Don't be shy. I'll be waiting.

import type { DndOpts, FinalDndOpts } from "../types";

export const defaultOpts: DndOpts = Object.freeze({
  threshold: {
    vertical: 60,
    horizontal: 60,
  },

  restrictions: {
    allowLeavingFromTop: true,
    allowLeavingFromBottom: true,
    allowLeavingFromLeft: true,
    allowLeavingFromRight: true,
  },

  scroll: {
    enable: false,
    speed: 10,
    threshold: {
      vertical: 75,
      horizontal: 75,
    },
  },
});

export function extractOpts(opts: DndOpts) {
  const options = { ...opts };

  (Object.keys(defaultOpts) as Array<keyof FinalDndOpts>).forEach((props) => {
    if (!options[props]) {
      options[props] = { ...defaultOpts[props] };

      return;
    }

    Object.keys(defaultOpts[props]).forEach((subProp) => {
      if (typeof defaultOpts[props][subProp] === "object") {
        if (options[props][subProp] !== undefined) {
          options[props][subProp] = {
            ...defaultOpts[props][subProp],
            ...options[props][subProp],
          };

          return;
        }

        options[props][subProp] = { ...defaultOpts[props][subProp] };

        return;
      }

      if (options[props][subProp] === undefined) {
        options[props][subProp] = defaultOpts[props][subProp];
      }
    });
  });

  return options as FinalDndOpts;
}

export default extractOpts;
