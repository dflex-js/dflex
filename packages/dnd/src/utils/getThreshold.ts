/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type { ThresholdPercentages, ThresholdInPixels } from "../types";

function getThreshold(
  thresholdInput: ThresholdPercentages,
  containerWidth: number,
  containerHeight: number
): ThresholdInPixels {
  const x = Math.round((thresholdInput.horizontal * containerWidth) / 100);

  const y = Math.round((thresholdInput.vertical * containerHeight) / 100);

  return {
    y,
    x,
  };
}

export default getThreshold;
