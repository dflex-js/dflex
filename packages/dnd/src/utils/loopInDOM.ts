function loopInDOM(
  fromElement: HTMLElement,
  // eslint-disable-next-line no-unused-vars
  cb: (arg: HTMLElement) => boolean
) {
  let current: HTMLElement | null = fromElement;

  do {
    if (cb(current)) {
      return current;
    }

    current = current.parentNode as HTMLElement;
  } while (current && !current.isSameNode(document.body));

  return null;
}

export default loopInDOM;
