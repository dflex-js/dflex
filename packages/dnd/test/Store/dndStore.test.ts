/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import store from "../../src/DnDStore";

import { elm1, elm2, elm3, elm4 } from "../utils";

describe("DnD Store", () => {
  beforeAll(() => {
    store.register(elm1);
    store.register(elm2);
    store.register(elm3);
    store.register(elm4);
  });

  it("Element is initiated", () => {
    expect(store.registry).toMatchSnapshot();
  });

  it("Returns all element connections and instance", () => {
    const elemInstance = store.getElmTreeById(elm1.id);

    expect(elemInstance).toMatchSnapshot();
  });
});
