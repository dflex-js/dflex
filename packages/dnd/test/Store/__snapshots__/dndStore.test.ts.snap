// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DnD Store Element is initiated 1`] = `
Object {
  "id-1": CoreInstance {
    "currentLeft": 450,
    "currentTop": 114,
    "hasToTransform": false,
    "id": "id-1",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "chK": null,
      "pK": "1-0",
      "sK": "0-0",
    },
    "offset": Object {
      "height": 50,
      "left": 450,
      "top": 114,
      "width": 170,
    },
    "order": Object {
      "parent": 0,
      "self": 0,
    },
    "prevTranslateY": Array [],
    "ref": <div
      data-index="0"
    />,
    "translateX": 0,
    "translateY": 0,
  },
  "id-2": CoreInstance {
    "currentLeft": 450,
    "currentTop": 172,
    "hasToTransform": false,
    "id": "id-2",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "chK": null,
      "pK": "1-0",
      "sK": "0-0",
    },
    "offset": Object {
      "height": 50,
      "left": 450,
      "top": 172,
      "width": 170,
    },
    "order": Object {
      "parent": 0,
      "self": 1,
    },
    "prevTranslateY": Array [],
    "ref": <div
      data-index="1"
    />,
    "translateX": 0,
    "translateY": 0,
  },
  "id-3": CoreInstance {
    "currentLeft": 450,
    "currentTop": 230,
    "hasToTransform": false,
    "id": "id-3",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "chK": null,
      "pK": "1-0",
      "sK": "0-0",
    },
    "offset": Object {
      "height": 50,
      "left": 450,
      "top": 230,
      "width": 170,
    },
    "order": Object {
      "parent": 0,
      "self": 2,
    },
    "prevTranslateY": Array [],
    "ref": <div
      data-index="2"
    />,
    "translateX": 0,
    "translateY": 0,
  },
  "id-4": CoreInstance {
    "currentLeft": 450,
    "currentTop": 288,
    "hasToTransform": false,
    "id": "id-4",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "chK": null,
      "pK": "1-0",
      "sK": "0-0",
    },
    "offset": Object {
      "height": 50,
      "left": 450,
      "top": 288,
      "width": 170,
    },
    "order": Object {
      "parent": 0,
      "self": 3,
    },
    "prevTranslateY": Array [],
    "ref": <div
      data-index="3"
    />,
    "translateX": 0,
    "translateY": 0,
  },
}
`;

exports[`DnD Store Returns all element connections and instance 1`] = `
Object {
  "branches": Object {
    "parents": undefined,
    "siblings": Array [
      "id-1",
      "id-2",
      "id-3",
      "id-4",
    ],
  },
  "element": CoreInstance {
    "currentLeft": 450,
    "currentTop": 114,
    "hasToTransform": false,
    "id": "id-1",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "chK": null,
      "pK": "1-0",
      "sK": "0-0",
    },
    "offset": Object {
      "height": 50,
      "left": 450,
      "top": 114,
      "width": 170,
    },
    "order": Object {
      "parent": 0,
      "self": 0,
    },
    "prevTranslateY": Array [],
    "ref": <div
      data-index="0"
    />,
    "translateX": 0,
    "translateY": 0,
  },
  "parent": null,
}
`;
